### PR TITLE
Hot Rod map storage: Authentication session no-downtime store

### LINKED ISSUE #9674: Hot Rod map storage: Authentication session no-downtime store
### Description

Create a Hot Rod map storage implementation for Authentication session area

### PR DESCRIPTION
**Pull request by martin-kanis**
Closes #9674

### CODE REVIEW

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/authSession/HotRodAuthenticationSessionEntity.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    public static abstract class AbstractHotRodAuthenticationSessionEntityDelegate extends UpdatableHotRodEntityDelegateImpl<HotRodAuthenticationSessionEntity> implements MapAuthenticationSessionEntity {
+
+        @Override
+        public Map<String, AuthenticationSessionModel.ExecutionStatus> getExecutionStatuses() {
+            Set<HotRodPair<String, HotRodExecutionStatus>> executionStatuses = getHotRodEntity().executionStatuses;
+            if (executionStatuses == null) {
+                return Collections.emptyMap();
+            }
+            return executionStatuses.stream().collect(Collectors.toMap(k -> k.getKey(),
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/authSession/HotRodAuthenticationSessionEntity.java:89]:
```suggestion
            return executionStatuses.stream().collect(Collectors.toMap(HotRodPair::getKey,
```

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/authSession/HotRodRootAuthenticationSessionEntity.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        @Override
+        public void clearUpdatedFlag() {
+            HotRodRootAuthenticationSessionEntity rootAuthSession = getHotRodEntity();
+            rootAuthSession.updated = false;
+            Optional.ofNullable(getAuthenticationSessions()).orElseGet(Collections::emptySet).forEach(UpdatableEntity::clearUpdatedFlag);
+        }
+
+        abstract public <V> V deepClone(V obj);
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/authSession/HotRodRootAuthenticationSessionEntity.java:107]: This seems to be unused

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/clientscope/HotRodClientScopeEntity.java [hunk 3]
```diff
@@ -28,7 +28,6 @@
 import org.keycloak.models.map.storage.hotRod.common.UpdatableHotRodEntityDelegateImpl;
 
 import java.util.Collection;
-import java.util.LinkedHashSet;
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/clientscope/HotRodClientScopeEntity.java:31]: This change seems to be unnecessary.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodTypesUtils.java [hunk 4]
```diff
@@ -110,4 +112,14 @@ public static String getKey(HotRodUserConsentEntity hotRodUserConsentEntity) {
     public static <T, V> Set<V> migrateSet(Set<T> p0, Function<T, V> migrator) {
         return p0 == null ? null : p0.stream().map(migrator).collect(Collectors.toSet());
     }
+
+    public static HotRodAuthenticationSessionEntity createHotRodAuthenticationSessionEntityFromMapEntry(Map.Entry<String, MapAuthenticationSessionEntity> entry) {
```
Comment (mhajas) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/common/HotRodTypesUtils.java:116]: This is unused after the change in MapAuthenticationSessionInterface (`Map` -> `Set`).

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you for the changes @martin-kanis. I added a few comments.

**Review by mhajas — DISMISSED** [review]
Thanks @martin-kanis

**Comment by hmlnarik** [comment]
@martin-kanis Is `ConcurrentLoginTest.concurrentLoginSingleUserSingleClient` from
[Base testsuite (undertow-map-hot-rod, group2)](https://github.com/keycloak/keycloak/runs/5571803540?check_suite_focus=true#logs) supposed to fail?

**Review by hmlnarik — APPROVED** [review]
Approving per @mhajas 's approval

**Comment by martin-kanis** [comment]
Model tests failure not related to this PR:
```
2022-03-21T19:47:40.9454832Z ##[error]| jpa-federation+infinispan | Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 66.379 s <<< FAILURE! - in org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest
2022-03-21T19:47:40.9466131Z ##[error]| jpa-federation+infinispan | org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.testPersistenceMultipleNodesClientSessionsAtRandomNode  Time elapsed: 10.953 s  <<< FAILURE!
2022-03-21T19:47:40.9468491Z ##[error]| jpa-federation+infinispan | java.lang.AssertionError: Execution didn't complete: java.lang.RuntimeException: org.infinispan.commons.CacheException: java.lang.IllegalStateException: Read commands must ignore leavers
```

**Pull request merged**